src/engine/MaxConstraint.h

```diff
@@ -34,137 +46,197 @@ class MaxConstraint : public PiecewiseLinearConstraint
+       The order of returned cases affects the search, and this method is where related
+       heuristics should be implemented.
+     */
+    List<PhaseStatus> getAllCases() const override;
+
+    /*
+       Checks whether a case has been flagged as infeasible. Works only in
+       context-dependent mode.
+     */
+    bool isCaseInfeasible( unsigned phase ) const;
+
+    /*
+      Check if the constraint's phase has been fixed.
     */
-    PiecewiseLinearCaseSplit getValidCaseSplit() const;
+    bool phaseFixed() const override;
 
     /*
       Return a list of smart fixes for violated constraint.
     */
-    List<PiecewiseLinearConstraint::Fix> getSmartFixes( ITableau *tableau ) const;
+    List<PiecewiseLinearConstraint::Fix> getSmartFixes( ITableau *tableau ) const override;
 
     /*
       Preprocessing related functions, to inform that a variable has been eliminated completely
       because it was fixed to some value, or that a variable's index has changed (e.g., x4 is now
       called x2). constraintObsolete() returns true iff and the constraint has become obsolote
       as a result of variable eliminations.
     */
-    void eliminateVariable( unsigned variable, double fixedValue );
-    void updateVariableIndex( unsigned oldIndex, unsigned newIndex );
-    bool constraintObsolete() const;
+    void eliminateVariable( unsigned variable, double fixedValue ) override;
+    void updateVariableIndex( unsigned oldIndex, unsigned newIndex ) override;
+    bool constraintObsolete() const override;
 
     /*
       Get the tightenings entailed by the constraint.
     */
-    void getEntailedTightenings( List<Tightening> &tightenings ) const;
+    void getEntailedTightenings( List<Tightening> &tightenings ) const override;
 
     /*
       For preprocessing: get any auxiliary equations that this constraint would
       like to add to the equation pool.
     */
-    void addAuxiliaryEquations( InputQuery &inputQuery );
+    void addAuxiliaryEquations( InputQuery &inputQuery ) override;
 
     /*
       Returns string with shape:
       max, _f, element_1, element_2, ... , element_n
     */
-    String serializeToString() const;
+    String serializeToString() const override;
 
     /*
      * Returns a boolean value indicating if at least one input variable was eliminated (True)
      * or not (False)
      */
     bool wereVariablesEliminated() const;
 
+
+
+    bool isImplication() const override;
+
 private:
     unsigned _f;
     Set<unsigned> _elements;
     Set<unsigned> _initialElements;
 
-    unsigned _maxIndex;
-    bool _maxIndexSet;
     double _maxLowerBound;
     bool _obsolete;
     bool _eliminatedVariables;
     double _maxValueOfEliminated;
 
+    /*
+       Functions that abstract away _maxIndex and _maxIndexSet, and use the
+       refactored PhaseStatus to store this information.
+
+       Makes two assumptions:
+        - 0 is not a valid variable value, since it's reserved for PHASE_NOT_FIXED
```

Comment:
0 is actually almost always a variable. can we get rid of this assumption?

Reply:
Yeah, I introduced the conversion functions since, it is simple to just increment the stored value by one when reading/writing.